src/test/fuzz/base_encode_decode.cpp

```diff
@@ -10,45 +10,86 @@
 {
-    const std::string random_encoded_string(buffer.begin(), buffer.end());
+    std::string random_string(buffer.begin(), buffer.end());
+    std::vector<unsigned char> random_data{ToByteVector(random_string)};
 
+    // Decode/Encode roundtrip
     std::vector<unsigned char> decoded;
-    if (DecodeBase58(random_encoded_string, decoded, 100)) {
-        const std::string encoded_string = EncodeBase58(decoded);
-        assert(encoded_string == TrimStringView(encoded_string));
-        assert(ToLower(encoded_string) == ToLower(TrimString(random_encoded_string)));
+    if (DecodeBase58(random_string, decoded, 100)) {
+        auto encoded_string{EncodeBase58(decoded)};
+        assert(encoded_string == TrimString(random_string));
+        assert(encoded_string.empty() || !DecodeBase58(encoded_string, decoded, decoded.size() - 1));
     }
+    // Encode/Decode roundtrip
+    auto encoded{EncodeBase58(random_data)};
+    std::vector<unsigned char> roundtrip_decoded;
+    assert(DecodeBase58(encoded, roundtrip_decoded, random_string.size())
+        && roundtrip_decoded == random_data);
+}
+
+FUZZ_TARGET(base58check_encode_decode)
+{
+    std::string random_string(buffer.begin(), buffer.end());
+    std::vector<unsigned char> random_data{ToByteVector(random_string)};
 
-    if (DecodeBase58Check(random_encoded_string, decoded, 100)) {
-        const std::string encoded_string = EncodeBase58Check(decoded);
-        assert(encoded_string == TrimString(encoded_string));
-        assert(ToLower(encoded_string) == ToLower(TrimString(random_encoded_string)));
+    // Decode/Encode roundtrip
+    std::vector<unsigned char> decoded;
+    if (DecodeBase58Check(random_string, decoded, 100)) {
+        auto encoded_string{EncodeBase58Check(decoded)};
+        assert(encoded_string == TrimString(random_string));
+        assert(encoded_string.empty() || !DecodeBase58Check(encoded_string, decoded, decoded.size() - 1));
     }
+    // Encode/Decode roundtrip
+    auto encoded{EncodeBase58Check(random_data)};
+    std::vector<unsigned char> roundtrip_decoded;
+    assert(DecodeBase58Check(encoded, roundtrip_decoded, random_string.size())
+        && roundtrip_decoded == random_data);
+}
+
+FUZZ_TARGET(base32_encode_decode)
+{
+    std::string random_string(buffer.begin(), buffer.end());
+    std::vector<unsigned char> random_data{ToByteVector(random_string)};
 
-    auto result = DecodeBase32(random_encoded_string);
-    if (result) {
-        const std::string encoded_string = EncodeBase32(*result);
-        assert(encoded_string == TrimStringView(encoded_string));
-        assert(ToLower(encoded_string) == ToLower(TrimString(random_encoded_string)));
+    // Decode/Encode roundtrip
+    if (auto result{DecodeBase32(random_string)}) {
+        auto encoded_string{EncodeBase32(*result)};
+        assert(encoded_string == ToLower(TrimString(random_string)));
     }
+    // Encode/Decode roundtrip
+    auto encoded{EncodeBase32(random_data)};
```

Comment:
Reusing the same data from the fuzz input for two different things (e.g. decode/encode and encode/decode) is usually not a good idea.

How about adding a decode step after the encoding above (i.e. decode -> encode -> decode)? Or alternatively split this second round-trip into its own harness as well?

Reply:
Thanks for the review!

> is usually not a good idea

Can you please expand on that? I don't mind splitting, as you've recommended, but I have to understand what the current drawbacks are.

Reply:
It might be less obvious (and maybe not even a problem) for this simple harness but the decode/encode and encode/decode round-trips are different tests. Their input spaces are not the same, which means that reusing the same fuzz input for both leads to less efficient fuzzing (e.g. the corpus will include inputs that are only relevant to one of the tests but both will be executed).

Reply:
Thanks @maflcko.
@dergoegge, would a `CallOneOf` based solution satisfy your needs?

Reply:
> There is also an overhead that we probably don't want a separate fuzz target for every trivial function, because each fuzz target may be put into a separate binary and is allocated a separate folder.

I don't think the binary sizes would be a problem after we merge something like https://github.com/bitcoin/bitcoin/pull/30882 and the extra folder for the corpus is also not a huge cost (they compress really nicely).

I'd prefer to follow best practices over having to weigh triviality against some external overhead of running these tests. "Put separate tests into separate harnesses" is easier to follow than "aggregate trivial functions in one harness" because it is entirely subjective. What is trivial? How many unrelated trivial functions are allowed in one harness? I think this creates confusion for devs that aren't experts on fuzzing.

> would a CallOneOf based solution satisfy your needs?

I don't think this solves the issue, e.g. mutations that change the calloneof branch result in completely changing the meaning of the fuzz input (i.e. switching between the different tests).

I don't mean to derail this PR, it's already improving things by splitting up `base_encode_decode`. Feel free to leave it as is.

Reply:
Thanks for the directions @dergoegge, @maflcko.

> I don't mean to derail this PR, it's already improving things by splitting up base_encode_decode. Feel free to leave it as is.

I'm all for separation of concerns, that's what I've done here as well.
If there's a simple and non-controvertial way to make it even better, let me know!

Reply:
(My instinct would be to go with the following to avoid rocking the boat when it comes to seed corpus naming etc, but I'm too new to this aspect of fuzzing to carry much weight in this aspect).
```C++
void base58_encode_decode(FuzzBufferType buffer);
void base58check_encode_decode(FuzzBufferType buffer);
void base32_encode_decode(FuzzBufferType buffer);
void base64_encode_decode(FuzzBufferType buffer);
void psbt_base64_decode(FuzzBufferType buffer);

FUZZ_TARGET(base_encode_decode)
{
    base58_encode_decode(buffer);
    base58check_encode_decode(buffer);
    base32_encode_decode(buffer);
    base64_encode_decode(buffer);
    psbt_base64_decode(buffer);
}
```

Reply:
Is this a blocker for anyone? If anyone has strong feelings about this, please let me know, we can either restructure here or provide it in another PR. Resolving for now.